posthog/models/cohort.py

```diff
@@ -102,41 +111,48 @@ def get_analytics_metadata(self):
             "deleted": self.deleted,
         }
 
-    def calculate_people(self):
+    def calculate_people(self, new_version: int, batch_size=50000, pg_batch_size=1000):
         if self.is_static:
             return
         try:
-            persons_query = self._clickhouse_persons_query()
 
-            try:
-                sql, params = persons_query.distinct("pk").only("pk").query.sql_with_params()
-            except EmptyResultSet:
-                query = DELETE_QUERY.format(cohort_id=self.pk)
-                params = {}
-            else:
-                query = f"""
-                    {DELETE_QUERY};
-                    {UPDATE_QUERY};
-                """.format(
-                    cohort_id=self.pk,
-                    values_query=sql.replace('FROM "posthog_person"', f', {self.pk} FROM "posthog_person"', 1,),
-                )
+            # Paginate fetch batch_size from clickhouse and paginate insert pg_batch_size into postgres
+            cursor = 0
+            persons = self._clickhouse_persons_query(batch_size=batch_size, offset=cursor)
+            while persons:
+                to_insert = [CohortPeople(person_id=p.pk, cohort_id=self.pk, version=new_version) for p in persons]
+                CohortPeople.objects.bulk_create(to_insert, batch_size=pg_batch_size)
+
+                cursor += batch_size
+                persons = self._clickhouse_persons_query(batch_size=batch_size, offset=cursor)
 
-            cursor = connection.cursor()
-            with transaction.atomic():
-                cursor.execute(query, params)
         except Exception as err:
+            # Clear the pending version people if there's an error
+            CohortPeople.objects.filter(cohort_id=self.pk, version=new_version).delete()
             raise err
 
-    def calculate_people_ch(self):
+    def calculate_people_ch(self, pending_version):
         from ee.clickhouse.models.cohort import recalculate_cohortpeople
-        from posthog.tasks.calculate_cohort import calculate_cohort
 
         try:
+            start_time = time.time()
```

Comment:
non-blocking

We should probably use [`monotonic`](https://docs.python.org/3/library/time.html#time.monotonic) here rather that `time`. `time` will be affected by clock changes.